integration_test/ops_agent_test.go

```diff
@@ -1893,6 +1901,181 @@ func TestPrometheusMetrics(t *testing.T) {
+		err := agents.InstallPackages(ctx, logger.ToMainLog(), vm, packages)
+		if err != nil {
+			t.Fatalf("failed to install %v with err: %s", packages, err)
+		}
+
+		// Run the setup script to run the Python http server and the JSON exporter
+		setupScript, err := os.ReadFile("testdata/setup_json_exporter.sh")
+		if err != nil {
+			t.Fatalf("failed to open setup script: %s", err)
+		}
+
+		env := map[string]string{"WORKDIR": workDir}
+		maxAttampts := 5
+		attempt := 0
+		for ; attempt < maxAttampts; attempt++ {
+			setupOut, err := gce.RunScriptRemotely(ctx, logger, vm, string(setupScript), nil, env)
+			// Since the script will start the processes in the background, the script should finish without error
+			if err != nil {
+				t.Fatalf("failed to run json exporter in VM with err: %v, stderr: %s", err, setupOut.Stderr)
+			}
+			// Wait until both are ready
+			time.Sleep(30 * time.Second)
+			liveCheckOut, err := gce.RunRemotely(ctx, logger.ToMainLog(), vm, "", `curl "http://localhost:7979/probe?module=default&target=http://localhost:8000/data.json"`)
+			// We will retry when:
+			// 1. JSON exporter is not started: in this case the stderr will have: "curl: (7) Failed to connect to localhost port 7979 after 1 ms: Connection refused"
+			// 2. The Python HTTP server is not started: in this case the stdout will not have the expected Prometheus style metrics
+			// If neither cases - break the retrying loop
+			if err == nil && !strings.Contains(liveCheckOut.Stderr, "Connection refused") && strings.Contains(liveCheckOut.Stdout, `test_counter_value{test_label="counter_label"} 1234`) {
+				break
+			}
+
+		}
+		if attempt == maxAttampts {
+			t.Fatalf("failed to start the HTTP server or JSON exporter - exhausted retries")
+		}
+
+		config := `metrics:
+  receivers:
+    prom_app:
+      type: prometheus
+      config:
+        scrape_configs:
+        - job_name: json
+          metrics_path: /probe
+          params:
+            module: [default]
+          static_configs:
+            - targets:
+              - http://localhost:8000/data.json 
+          relabel_configs:
+            - source_labels: [__address__]
+              target_label: __param_target
+              replacement: '$1'
+            - source_labels: [__param_target]
+              target_label: instance
+              replacement: '$1'
+            - target_label: __address__
+              replacement: localhost:7979 
+  service:
+    pipelines:
+      prom_pipeline:
+        receivers: [prom_app]
+`
+
+		if err := setupOpsAgent(ctx, logger, vm, config); err != nil {
+			t.Fatal(err)
+		}
+
+		// Wait long enough for the data to percolate through the backends
+		// under normal circumstances. Based on some experiments, 2 minutes
+		// is normal; wait a bit longer to be on the safe side.
+		time.Sleep(3 * time.Minute)
+		window := time.Minute
+
+		var multiErr error
+		// Gauge type
+		gaugeMetric := "prometheus.googleapis.com/test_gauge_value/gauge"
```

Comment:
A lot of the tests look like they can be templated. Something like:

```
type promMetricTest struct {
  metricName string
  metricKind   MetricKind
  valueType     ValueType
  expectedValue float64
}

testData = []promMetricTest{
{"prometheus.googleapis.com/test_gauge_value/gauge", metric.MetricDescriptor_GAUGE, metric.MetricDescriptor_DOUBLE, 789},
{"prometheus.googleapis.com/test_counter_value/counter", metric.MetricDescriptor_CUMULATIVE, metric.MetricDescriptor_DOUBLE, 0},
{"prometheus.googleapis.com/test_untyped_metric_value/gauge", metric.MetricDescriptor_GAUGE, metric.MetricDescriptor_DOUBLE, 56},
}

for test := range testData {
  ...
}
```

Reply:
Great suggestion! I have made the change to reuse code for all the metric types.